nixos/lib/test-driver/test_driver/driver.py

```diff
@@ -161,6 +172,12 @@ def join_all(self) -> None:
         with rootlog.nested("wait for all VMs to finish"):
             for machine in self.machines:
                 machine.wait_for_shutdown()
+            self.race_timer.cancel()
+
+    def terminate_test(self) -> None:
+        with rootlog.nested("timeout reached; test termination..."):
+            for machine in self.machines:
+                machine.release()
```

Comment:
This doesn't seem to handle non-termination in the `testScript`.
Also what happens if the `testScript` has a `try`/`except`? The option doc claims that the script will fail, but this does not seem to be guaranteed.

Reply:
What do you mean exactly by non-termination? The test passes? Termination is done somewhere else (by releasing all the machines too with a different message).

Good point I guess about the `try/except`, I wonder if there's any form of non-catchable exception inside Python but everyone would break down in general.

In any case, I guess we could ensure termination by exiting with exit code 2 (to make it clear it's a timeout-driven exit).

Reply:
Here's a proposal, I will kill the main thread in the termination sequence.

Reply:
(more precisely: I will kill the process and let the kernel do the right thing.)

Reply:
> kill the process

sgtm.

Alternatively, python may have a direct exit "syscall" that doesn't throw an exception but just does the thing immediately.

Reply:
This seems to be it: [os._exit](https://docs.python.org/3/library/os.html#os._exit)

Reply:
Oh right, I guess using `kill` is less obscure than `_exit`, if that's OK with you.

Reply:
Not sure if that's so great though. `sys.exit` might throw an exception and therefore not be airtight, but you don't get any cleanup either. Script author won't be able to clean up out-of-process resources, such as child processes.
Maybe it should just be `sys.exit` anyway, although I don't know how that exception would propagate to the main thread?

Reply:
Usually we have Nix to fall back on as a kind of process manager, but that's not the case with interactive use (`nix run .#foo.driverInteractive`).

Reply:
For non-interactive usecase, this is exactly what we want, I would say.
For interactive usecase, well… it was ruled out of scope from the get-go and I would argue that timeout should be disabled in the interactive usecase.

Using `sys.exit` is impossible because there's no way we can propagate the `threading.Timer` exception back to the main thread.

We would need our own way to control error handling and offer a chance to terminate gracefully, I would expect this would require some serious back'n'forth in the code to achieve this, including running the test script on its own thread and having the main thread as a manager.

AFAIK, out of process resources should still be killed in interactive because they are spawned off from a unique hierarchy (the interactive runner).

So… I would argue this is as good as we can go without investing much more time, which I am afraid I won't be able to commit to this in particular.

Reply:
> there's no way we can propagate

Makes sense actually. Usually only languages that exercise tight control over their runtime allow this, examples being Java and Haskell.
(Also pthreads I guess, but unsafely)

Sounds good enough to me then.